### DISCUSSION TITLE
Potential Redundancy Between TanStack Router and Query Features?

### DISCUSSION THREAD

**Discussion by eastack** [TanStack/router#830] [Q&A] [answered]
Do TanStack Router's Loader and Actions have overlapping functionality with TanStack Query's Queries and Mutations? If my understanding is incorrect, are there any examples combining the features of both?

**Comment by eastack** [comment]
For example: `queryClient.invalidateQueries(...)` and `loaderClient.invalidateLoader(...)`. I feel very confused about this. How should I use both together or is using TanStack Router enough?

**Comment by eastack** [comment, marked as answer]
I think I understand now. Loaders and Actions no longer exist, which is why I was wondering previously why their versions were stuck at 0.0.1-beta.204. It is now recommended to use TanStack Query instead.

![image](https://github.com/TanStack/router/assets/17524858/b8b38d9f-9c46-4d98-916c-2f5f03ab21bf)
![image](https://github.com/TanStack/router/assets/17524858/14bc0ded-7837-4c28-8146-f8dc309d1104)

**Comment by tannerlinsley** [comment]
The router still has very basic loader functionality built in. It should handle many use cases, but also integrate well with Query when you need it.

**Reply by eastack** [reply]
Sorry I have another question. I'm now developing a side project using TanStack Router and TanStack Query. I want to utilize React's suspense functionality, but I found that using `queryClient.ensureQueryData` directly in the `loader` and then using `useSuspenseQuery` in components, this leads to routes being blocked and the page keeps showing blank.
(If I have done anything foolish, please forgive me as I am a front-end novice. I'm currently only trying to find some information in the examples and docs.)

I found that the previous examples don't seem to apply anymore. In one example, the function named `defer` and the `Await` component have been removed.

![image](https://github.com/TanStack/router/assets/17524858/255f0497-fb9e-4b02-a0b7-16efe3def70b)
![image](https://github.com/TanStack/router/assets/17524858/b26121a6-335a-47a5-bb24-e409c57cc59d)

**Reply by eastack** [reply]
I think I roughly understand now. If I directly return a Promise, TanStack Router will try to wait for it to complete or fail. Seeing this code snippet in the issue gave me inspiration. Now everything works fine. The whole process is like solving a puzzle, haha.

![image](https://github.com/TanStack/router/assets/17524858/0aff1b42-f920-41b5-bf99-a7b68e2eec4a)

**Reply by boldurean** [reply]
This more seems like a hack rather than solution :|

**Reply by TkDodo** [reply]
if you return a promise from the loader, the router will await it and will show the `pendingComponent` for your route. If you don't have a `pendingComponent` specified, you won't see any loader, so that's expected.

It's also fine to use the router to kick off the loading by _not_ returning the promise as the example shows. You could also do:

```
loader: () => void queryClient.ensureQueryData(...)
```

then, in your component, you can `useSuspenseQuery`, but if you don't define your own suspense boundaries, it will pick the one from the router, so again the `pendingComponent` will be shown.

**Reply by boldurean** [reply]
I've found ideal solution for me. The Route loader ensuresQueryData, and component still fetches with useQuery using same queryOptions, which makes possible preload/or not if data exists on hover over the Link. Benefits: no changes in component while refactoring, no additional Suspense wrappers while moving from the react-router-dom :)

**Reply by TkDodo** [reply]
this is literally what the router docs suggest 😅 

https://tanstack.com/router/latest/docs/framework/react/guide/external-data-loading#a-more-realistic-example-using-tanstack-query

**Reply by boldurean** [reply]
@TkDodo  docs suggest to useSuspenseQuery :) i'm using loader to prefech query, and useQuery inside component :P

**Reply by TkDodo** [reply]
Why not suspenseQuery? Integrates natively with suspense boundaries from the router, and ensures data can never be undefined on type level

**Reply by boldurean** [reply]
The query takes 2-3 seconds to be processed and here is just one Suspense component at the top level. If I will use this it will trigger suspense and no other components will be rendered while loading.
We decided for smoother migration over the tanstack router just to use existing queries, and later on add inner Suspense + replace components useQuery -> useSuspenseQuery.
